### PR TITLE
[cc] [refactor] Automatically detect available compiler (clang or gcc)

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1644

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
We will choose to `clang` by default, `gcc` as fallback.

mpm88:
cc gcc: 18 fps
cc clang: 24 fps
opengl mesa: 25 fps
Amazing! We're almost beating OpenGL on single core CPU! Maybe this's the power of clang's SIMD optimizer?

I also noticed that env option `TI_CC_COMPILER_CMD=xxx` doesn't work, will fix iapr.
Btw, @k-ye would you help me setup the C backend on OS X?

### CODE REVIEW

File: taichi/backends/cc/codegen_cc.cpp [hunk 1]
```diff
@@ -470,18 +471,7 @@ class CCTransformer : public IRVisitor {
...
-        emit("Ti_i32 {} = {};", stmt->raw_name(), stmt->loop->raw_name());
-      } else {
-        TI_NOT_IMPLEMENTED
-      }
-    } else if (stmt->loop->is<RangeForStmt>()) {
-      emit("Ti_i32 {} = {};", stmt->raw_name(), stmt->loop->raw_name());
-    } else {
-      TI_NOT_IMPLEMENTED
-    }
+    emit("Ti_i32 {} = {};", stmt->raw_name(), stmt->loop->raw_name());
```
Comment (xumingkuan) [inline taichi/backends/cc/codegen_cc.cpp:474]: I suggest reverting this change -- IIUC, we should fail loudly if there are struct fors.

**Review by xumingkuan — COMMENTED** [review]
LGTMig! I'm not sure if clang is better than gcc in other cases though.

File: taichi/backends/cc/codegen_cc.cpp [hunk 2]
```diff
@@ -329,18 +329,16 @@ class CCTransformer : public IRVisitor {
       if (is_comparison(bin->op_type)) {
         // XXX(#577): Taichi uses -1 as true due to LLVM i1...
         emit("{} = -({} {} {});", var, lhs_name, binop, rhs_name);
+
       } else if (bin->op_type == BinaryOpType::truediv) {
         emit("{} = ({}) {} / {};", var, dt_name, lhs_name, rhs_name);
+
       } else if (bin->op_type == BinaryOpType::floordiv) {
+        TI_WARN(
+            "floordiv called! It should be taken care by demote_operations");
```
Comment (xumingkuan) [inline taichi/backends/cc/codegen_cc.cpp:338]:
```suggestion
            "floordiv called! It should be taken care of by demote_operations");
```

**Review by xumingkuan — APPROVED** [review]

**Comment by archibate** [comment]
> LGTMig! I'm not sure if clang is better than gcc in other cases though.

Yes, tree_gravity seems slower with clang. Maybe gcc is better in serial works while clang is good at simd? Anyway as most usages of taichi are parallelization, imo simd can be beneficial to cc.

**Comment by yuanming-hu** [comment]
Sorry but I have to revert this PR, since it seems to crash on Jenkins (`gcc`):
(Feel free to re-open when it's fixed.)

```
Running test on arch=Arch.cc
[Taichi] Starting on arch=cc
[W 11/30/20 11:17:41.241] [cc_program.cpp:CCProgram@178] 
[I 11/30/20 11:17:41.304] [cc_program.cpp:smart_choose_compiler@165] [cc] using `gcc -Wc99-c11-compat -Wc11-c2x-compat` as compiler
[Taichi] materializing...
[E 11/30/20 11:17:41.562] [cc_program.cpp:compile@75] [cc] could not load shared object: /tmp/taichi-oiudtua3/libti_roottest.so


[38;2;255;000;255m***********************************
[0m[38;2;255;000;255m* Taichi Compiler Stack Traceback *
[0m[38;2;255;000;255m***********************************
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so: taichi::Logger::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so: taichi::lang::cccp::CCLayout::compile()
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so: taichi::lang::cccp::CCProgram::compile_layout(taichi::lang::SNode*)
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so: taichi::lang::Program::materialize_layout()
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so: taichi::lang::layout(std::function<void ()> const&)
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so(+0x560f49) [0x7f4ee2bc1f49]
[0m[38;2;255;000;255m/tmp/taichi-oiudtua3/taichi_core.so(+0x3e5f13) [0x7f4ee2a46f13]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a4a5]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588d41]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5576d8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x541a90]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588e5c]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588e5c]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588e5c]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588c8b]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x54a971]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyObject_FastCallKeywords
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a433]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588c8b]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x54a971]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588d41]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588c8b]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x54a971]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyObject_FastCallKeywords
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a433]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588c8b]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x54a971]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyObject_FastCallKeywords
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a433]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588c8b]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x54a971]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyObject_FastCallKeywords
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a433]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x5095c8]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x509900]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x50a2fd]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x588d41]
[0m[38;2;255;000;255m/usr/bin/python3.6: PyObject_Call
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyEval_EvalFrameDefault
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x507be4]
[0m[38;2;255;000;255m/usr/bin/python3.6: _PyFunction_FastCallDict
[0m[38;2;255;000;255m/usr/bin/python3.6: ) [0x594a01]
[0m[38;2;255;000;255m/usr/bin/python3.6: )
----------------------------- Captured stderr call -----------------------------
gcc: error: unrecognized command line option ‘-Wc11-c2x-compat’; did you mean ‘-Wc90-c99-compat’?
gcc: error: /tmp/taichi-oiudtua3/_rti_roottest.o: No such file or directory
gcc: error: unrecognized command line option ‘-Wc11-c2x-compat’; did you mean ‘-Wc90-c99-compat’?
gcc: fatal error: no input files
compilation terminated.
____________________________ test_overflow[dt5-32] _____________________________

dt = <taichi_core.DataType object at 0x7f4ee10d80a0>, n = 32

    @pytest.mark.parametrize('dt,n', [
        (ti.i8, 8),
        (ti.u8, 8),
        (ti.i16, 16),
        (ti.u16, 16),
        (ti.i32, 32),
        (ti.u32, 32),
    ])
    @ti.archs_excluding(ti.opengl)
    def test_overflow(dt, n):
>       _test_overflow(dt, n)
```

**Comment by archibate** [comment]
Oh I see, seems the old version of gcc doesn't have the c2x option.

**Pull request merged**